From 471d68d93903af04f25b7b7bffc1872af74329fe Mon Sep 17 00:00:00 2001
From: Krisztian Litkey <krisztian.litkey@intel.com>
Date: Thu, 17 Dec 2015 13:06:17 +0200
Subject: [PATCH 2/2] nspawn: added support for starting DHCP client in
 non-system containers.

Added --auto-dhcp option for starting a DHCP client in the container
before privileges are dropped or the actual container command gets
executed. This is done only for non-system containers started without
--share-system.

The DHCP client command can be specified as an argument to the option.
If none is specified {/sbin,/usr/sbin}/{dhclient,udhcpc} are tried in
this particular order. The network interface name is assumed to be
'host0'.
---
 src/nspawn/nspawn.c | 106 ++++++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 106 insertions(+)

diff --git a/src/nspawn/nspawn.c b/src/nspawn/nspawn.c
index e0c90f8..54491f7 100644
--- a/src/nspawn/nspawn.c
+++ b/src/nspawn/nspawn.c
@@ -152,6 +152,8 @@ static const char *arg_selinux_context = NULL;
 static const char *arg_selinux_apifs_context = NULL;
 static const char *arg_slice = NULL;
 static bool arg_private_network = false;
+static char *arg_dhcp_client = NULL;
+static bool arg_auto_dhcp = false;
 static bool arg_read_only = false;
 static bool arg_boot = false;
 static bool arg_ephemeral = false;
@@ -413,6 +415,7 @@ static int parse_argv(int argc, char *argv[]) {
                 ARG_PROPERTY,
                 ARG_PRIVATE_USERS,
                 ARG_KILL_SIGNAL,
+                ARG_AUTO_DHCP,
         };
 
         static const struct option options[] = {
@@ -455,6 +458,7 @@ static int parse_argv(int argc, char *argv[]) {
                 { "property",              required_argument, NULL, ARG_PROPERTY          },
                 { "private-users",         optional_argument, NULL, ARG_PRIVATE_USERS     },
                 { "kill-signal",           required_argument, NULL, ARG_KILL_SIGNAL       },
+                { "auto-dhcp",             optional_argument, NULL, ARG_AUTO_DHCP         },
                 {}
         };
 
@@ -957,6 +961,11 @@ static int parse_argv(int argc, char *argv[]) {
 
                         break;
 
+                case ARG_AUTO_DHCP:
+                        arg_auto_dhcp = true;
+                        arg_dhcp_client = optarg;
+                        break;
+
                 case '?':
                         return -EINVAL;
 
@@ -1017,6 +1026,11 @@ static int parse_argv(int argc, char *argv[]) {
                 return -EINVAL;
         }
 
+        if (arg_auto_dhcp && (arg_boot || arg_share_system)) {
+                log_error("--dhcp-client may not be combined with --boot or --share-system.");
+                return -EINVAL;
+        }
+
         if (arg_userns && access("/proc/self/uid_map", F_OK) < 0)
                 return log_error_errno(EOPNOTSUPP, "--private-users= is not supported, kernel compiled without user namespace support.");
 
@@ -3696,6 +3710,93 @@ static int spawn_getent(const char *database, const char *key, pid_t *rpid) {
         return pipe_fds[0];
 }
 
+
+static int spawn_dhcp_client(pid_t *rpid) {
+        pid_t pid;
+        const char *interface = "host0";
+        const char *argv[16];
+
+        assert(rpid);
+
+        pid = fork();
+        if (pid < 0)
+                return log_error_errno(errno, "Failed to fork dhclient: %m");
+        else if (pid == 0) {
+                int nullfd;
+                char *empty_env = NULL;
+
+                nullfd = open("/dev/null", O_RDWR);
+                if (nullfd < 0)
+                        _exit(EXIT_FAILURE);
+
+                if (dup3(nullfd, STDIN_FILENO, 0) < 0)
+                        _exit(EXIT_FAILURE);
+
+                if (dup3(nullfd, STDOUT_FILENO, 0) < 0)
+                        _exit(EXIT_FAILURE);
+
+                if (dup3(nullfd, STDERR_FILENO, 0) < 0)
+                        _exit(EXIT_FAILURE);
+
+                if (nullfd > 2)
+                        safe_close(nullfd);
+
+                (void) reset_all_signal_handlers();
+                (void) reset_signal_mask();
+                close_all_fds(NULL, 0);
+
+                if (arg_dhcp_client) {
+                        char **args = strv_split(arg_dhcp_client, " ");
+                        char **a;
+                        int i;
+
+                        if (args == NULL)
+                                _exit(EXIT_FAILURE);
+
+                        i = 0;
+                        STRV_FOREACH(a, args) {
+                                if (i >= (int)ELEMENTSOF(argv) - 1)
+                                        _exit(EXIT_FAILURE);
+
+                                if (i == 0) {
+                                        char *p = strrchr(*a, '/');
+
+                                        argv[i] = p ? p + 1 : *a;
+                                }
+                                else
+                                        argv[i] = *a;
+                        }
+                        argv[i] = interface;
+
+                        execve(args[0], (char *const *)argv, &empty_env);
+                        _exit(EXIT_FAILURE);
+                }
+                else {
+                        argv[0] = "dhclient";
+                        argv[1] = interface;
+                        argv[2] = NULL;
+
+                        execve("/usr/sbin/dhclient", (char *const*)argv, &empty_env);
+                        execve("/sbin/dhclient", (char *const*)argv, &empty_env);
+
+                        argv[0] = "udhcpc";
+                        argv[1] = "-i";
+                        argv[2] = interface;
+                        argv[3] = NULL;
+
+                        execve("/usr/sbin/udhcpc", (char *const*)argv, &empty_env);
+                        execve("/sbin/udhcpc", (char *const*)argv, &empty_env);
+
+                        _exit(EXIT_FAILURE);
+                }
+        }
+
+        *rpid = pid;
+
+        return 0;
+}
+
+
 static int change_uid_gid(char **_home) {
         char line[LINE_MAX], *x, *u, *g, *h;
         const char *word, *state;
@@ -4090,6 +4191,7 @@ static int inner_child(
                 NULL, /* LISTEN_PID */
                 NULL
         };
+        pid_t pid;
 
         _cleanup_strv_free_ char **env_use = NULL;
         int r;
@@ -4150,6 +4252,10 @@ static int inner_child(
                 return r;
         rtnl_socket = safe_close(rtnl_socket);
 
+        if (arg_private_network && arg_auto_dhcp)
+                if (spawn_dhcp_client(&pid) < 0)
+                        return log_error_errno(errno, "spawn_dhcp_client() failed: %m");
+
         if (drop_capabilities() < 0)
                 return log_error_errno(errno, "drop_capabilities() failed: %m");
 
-- 
2.5.0

